File: html/js/pbclass/PbFile.js
// PbFile.js — port of Java PbFile.java (browser-compatible, no disk I/O)
// Receives file data as Uint8Array (pre-extracted from ZIP by the caller).
import * as BH from '../BufferHelper.js';
import { PbEntry } from './PbEntry.js';

const _ascii = new TextDecoder('ascii');
const _utf16 = new TextDecoder('utf-16le');

function _str(data, offset, len, unicode = false) {
    const slice = data.slice(offset, offset + len);
    return unicode ? _utf16.decode(slice) : _ascii.decode(slice);
}

export class PbFile {
    /**
     * Normal constructor: parse a PBD/PBL from in-memory Uint8Array.
     * @param {import('./PbProject.js').PbProject} project
     * @param {string} fileName — display name (no path needed)
     * @param {Uint8Array} data — full file content
     */
    constructor(project, fileName, data) {
        this.project  = project;
        this.fileName = fileName;
        this.filePath = fileName;
        /** @type {PbEntry[]} */
        this.entries  = [];

        const nodes = this._getNodeList(data);
        for (const node of nodes) {
            let current = node + 32;
            const buf32 = data.slice(node, node + 32);
            const uShort  = BH.getUShort(buf32, 20);
            const bytesPer = !project.isUnicode ? 1 : 2;
            const nameHdrSize = 4 + bytesPer * 4;
            const num3  = nameHdrSize + 16;
            const array = new Uint8Array(num3);

            for (let i = 0; i < uShort; i++) {
                array.set(data.slice(current, current + num3));
                const tag4 = _ascii.decode(array.slice(0, 4));
                const ver4 = project.getString(array, 4, bytesPer * 4);
                if (tag4 !== 'ENT*' || (ver4 !== '0600' && ver4 !== '0500')) {
                    throw new Error('格式错误');
                }
                const uInt  = BH.getUInt(array, nameHdrSize);
                const uInt2 = BH.getUInt(array, nameHdrSize + 4);
                const uShort2 = BH.getUShort(array, nameHdrSize + 14);
                current += num3;

                const buffer2 = data.slice(current, current + uShort2);
                current += uShort2;

                const entryName = project.getString(buffer2, 0, uShort2 - bytesPer);
                this.entries.push(new PbEntry(this, entryName, PbFile._readData(data, uInt, uInt2)));
            }
        }
    }

    /**
     * System-library constructor. In the browser we skip loading embedded
     * resources — just return an empty PbFile-like object.
     * @param {import('./PbProject.js').PbProject} project
     * @param {number} version
     * @returns {PbFile}
     */
    static createSystem(project, version) {
        const f = Object.create(PbFile.prototype);
        f.project  = project;
        f.fileName = 'system';
        f.filePath = null;
        f.entries  = [];
        // In the browser we cannot load embedded gzip resources — skip silently.
        return f;
    }

    // ── Private helpers ──────────────────────────────────────────────────────

    /**
     * @param {Uint8Array} data
     * @param {number} start   — offset of first DAT* block
     * @param {number} size    — total expected bytes
     * @returns {Uint8Array}
     */
    static _readData(data, start, size) {
        const out  = new Uint8Array(size);
        let num    = 0;
        let offset = start;
        const hdr  = new Uint8Array(10);
        while (num < size) {
            hdr.set(data.slice(offset, offset + 10));
            if (_ascii.decode(hdr.slice(0, 4)) !== 'DAT*') break;
            const chunkSize = BH.getUShort(hdr, 8);
            out.set(data.slice(offset + 10, offset + 10 + chunkSize), num);
            num    += chunkSize;
            offset  = BH.getUInt(hdr, 4);
        }
        return out;
    }

    /**
     * Walk NOD* blocks and return their offsets.
     * @param {Uint8Array} data
     * @returns {number[]}
     */
    _getNodeList(data) {
        const list = [];
        const array = new Uint8Array(512);
        let foundAt = -1;
        let pos = 0;

        // Scan for HDR*PowerBuilder header
        while (pos + 512 <= data.length) {
            array.set(data.slice(pos, pos + 512));
            const tag4 = _ascii.decode(array.slice(0, 4));
            if (tag4 === 'HDR*') {
                const pb12 = _ascii.decode(array.slice(4, 16));
                if (pb12 === 'PowerBuilder') {
                    const ver4 = _ascii.decode(array.slice(18, 22));
                    if (ver4 === '0500') {
                        this.project.isPb5    = true;
                        foundAt = pos;
                        break;
                    }
                    if (ver4 === '0600') {
                        foundAt = pos;
                        break;
                    }
                }
                // try unicode header
                const pbU = _utf16.decode(array.slice(4, 28));
                const verU = _utf16.decode(array.slice(32, 40));
                if (pbU === 'PowerBuilder' && verU === '0600') {
                    this.project.isUnicode = true;
                    foundAt = pos;
                    break;
                }
            }
            pos += 512;
        }

        if (foundAt < 0) return list;

        // Skip past the HDR block
        let num = foundAt + (this.project.isUnicode ? 1536 : 1024);
        if (num + 512 > data.length) throw new Error('格式错误');
        array.set(data.slice(num, num + 512));
        if (_ascii.decode(array.slice(0, 4)) !== 'NOD*') throw new Error('格式错误');

        list.push(num);
        let uInt  = BH.getUInt(array, 4);
        let uInt2 = BH.getUInt(array, 12);

        // Forward chain
        while (uInt !== 0) {
            array.set(data.slice(uInt, uInt + 512));
            if (_ascii.decode(array.slice(0, 4)) !== 'NOD*') throw new Error('格式错误');
            list.push(uInt);
            uInt = BH.getUInt(array, 4);
        }
        // Backward chain
        while (uInt2 !== 0) {
            array.set(data.slice(uInt2, uInt2 + 512));
            if (_ascii.decode(array.slice(0, 4)) !== 'NOD*') throw new Error('格式错误');
            list.push(uInt2);
            uInt2 = BH.getUInt(array, 12);
        }

        return list;
    }
}

File: java/src/main/java/com/pbdviewer/utils/pbclass/PbFile.java
package com.pbdviewer.utils.pbclass;

import com.pbdviewer.utils.BufferHelper;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.RandomAccessFile;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.GZIPInputStream;

public class PbFile {

    private final PbProject project;
    private final String filePath;
    private final String fileName;
    private final List<PbEntry> entries = new ArrayList<>();

    // Constructor for real file
    public PbFile(PbProject project, String filePath) {
        this.project = project;
        this.filePath = filePath;
        String[] parts = filePath.replace('\\', '/').split("/");
        this.fileName = parts[parts.length - 1];

        try (RandomAccessFile raf = new RandomAccessFile(filePath, "r")) {
            for (long node : getNodeList(raf)) {
                byte[] buffer = new byte[32];
                raf.seek(node);
                raf.read(buffer, 0, 32);
                long current = raf.getFilePointer();
                int uShort = BufferHelper.getUShort(buffer, 20L);
                int num = (!project.isUnicode()) ? 1 : 2;
                int num2 = 4 + num * 4;
                int num3 = num2 + 16;
                byte[] array = new byte[num3];
                for (int i = 0; i < uShort; i++) {
                    raf.seek(current);
                    raf.read(array, 0, num3);
                    if (!new String(array, 0, 4, StandardCharsets.US_ASCII).equals("ENT*") ||
                        (!project.getString(array, 4, num * 4).equals("0600") &&
                         !project.getString(array, 4, num * 4).equals("0500"))) {
                        throw new RuntimeException("格式错误");
                    }
                    long uInt = BufferHelper.getUInt(array, num2);
                    long uInt2 = BufferHelper.getUInt(array, num2 + 4);
                    int uShort2 = BufferHelper.getUShort(array, num2 + 14);
                    byte[] buffer2 = new byte[uShort2];
                    raf.read(buffer2, 0, uShort2);
                    current = raf.getFilePointer();
                    String entryName = project.getString(buffer2, 0, uShort2 - num);
                    entries.add(new PbEntry(this, entryName, readData(raf, uInt, uInt2)));
                }
            }
        } catch (IOException e) {
            throw new RuntimeException("Failed to read PBD file: " + filePath, e);
        }
    }

    // Constructor for system library (from embedded resource)
    public PbFile(PbProject project, int version) {
        this.project = project;
        this.filePath = null;
        this.fileName = "system";

        String resourcePath = String.format("/resoures/%04x.bin", version);
        InputStream stream = PbFile.class.getResourceAsStream(resourcePath);
        if (stream != null) {
            try {
                byte[] array = new byte[4096];
                GZIPInputStream gzipStream = new GZIPInputStream(stream);
                ByteArrayOutputStream memoryStream = new ByteArrayOutputStream();
                int num;
                do {
                    num = gzipStream.read(array, 0, array.length);
                    if (num > 0) memoryStream.write(array, 0, num);
                } while (num > 0);
                gzipStream.close();
                stream.close();
                byte[] entryData = memoryStream.toByteArray();
                memoryStream.close();
                entries.add(new PbEntry(this, "_typedef.grp", entryData));
            } catch (IOException e) {
                throw new RuntimeException("Failed to read system library resource", e);
            }
        }
    }

    private static byte[] readData(RandomAccessFile raf, long start, long size) throws IOException {
        byte[] array = new byte[(int) size];
        int num = 0;
        long offset = start;
        byte[] array2 = new byte[10];
        while (num < size) {
            raf.seek(offset);
            raf.read(array2, 0, 10);
            if (!new String(array2, 0, 4, StandardCharsets.US_ASCII).equals("DAT*")) break;
            int uShort = BufferHelper.getUShort(array2, 8L);
            raf.read(array, num, uShort);
            num += uShort;
            offset = BufferHelper.getUInt(array2, 4L);
        }
        return array;
    }

    private List<Long> getNodeList(RandomAccessFile raf) throws IOException {
        List<Long> list = new ArrayList<>();
        byte[] array = new byte[512];
        long num = 0L;
        boolean flag = false;
        raf.seek(0L);
        int num2 = raf.read(array, 0, 512);
        while (num2 > 0) {
            if (new String(array, 0, 4, StandardCharsets.US_ASCII).equals("HDR*")) {
                if (new String(array, 4, 12, StandardCharsets.US_ASCII).equals("PowerBuilder")) {
                    if (new String(array, 18, 4, StandardCharsets.US_ASCII).equals("0500")) {
                        project.setPb5(true);
                        flag = true;
                        break;
                    }
                    if (new String(array, 18, 4, StandardCharsets.US_ASCII).equals("0600")) {
                        flag = true;
                        break;
                    }
                }
                if (new String(array, 4, 24, StandardCharsets.UTF_16LE).equals("PowerBuilder") &&
                    new String(array, 32, 8, StandardCharsets.UTF_16LE).equals("0600")) {
                    flag = true;
                    project.setUnicode(true);
                    break;
                }
            }
            num += num2;
            num2 = raf.read(array, 0, 512);
        }
        if (flag) {
            num += (project.isUnicode() ? 1536 : 1024);
            raf.seek(num);
            num2 = raf.read(array, 0, 512);
            if (num2 != 512 || !new String(array, 0, 4, StandardCharsets.US_ASCII).equals("NOD*")) {
                throw new RuntimeException("格式错误");
            }
            list.add(num);
            long uInt = BufferHelper.getUInt(array, 4L);
            long uInt2 = BufferHelper.getUInt(array, 12L);
            while (uInt != 0) {
                raf.seek(uInt);
                num2 = raf.read(array, 0, 512);
                if (num2 != 512 || !new String(array, 0, 4, StandardCharsets.US_ASCII).equals("NOD*")) {
                    throw new RuntimeException("格式错误");
                }
                list.add(uInt);
                uInt = BufferHelper.getUInt(array, 4L);
            }
            while (uInt2 != 0) {
                raf.seek(uInt2);
                num2 = raf.read(array, 0, 512);
                if (num2 != 512 || !new String(array, 0, 4, StandardCharsets.US_ASCII).equals("NOD*")) {
                    throw new RuntimeException("格式错误");
                }
                list.add(uInt2);
                uInt2 = BufferHelper.getUInt(array, 12L);
            }
        }
        return list;
    }

    // -------------------------------------------------------------------------
    // Getters
    // -------------------------------------------------------------------------

    public PbProject getProject() {
        return project;
    }

    public String getFilePath() {
        return filePath;
    }

    public String getFileName() {
        return fileName;
    }

    public List<PbEntry> getEntries() {
        return entries;
    }
}
